File: Node/src/services/PostAnalyticsService.ts
import DatabaseConnection from '../database/connection';
import { Database } from 'sqlite3';
import { Post } from '../models/Post';

// ❌ PROBLEMA: Singleton mal implementado - vai criar problema de concorrência
export class PostAnalyticsService {
  private static instance: PostAnalyticsService;
  private db: Database;
  private cache: Map<string, any> = new Map();

  private constructor() {
    this.db = DatabaseConnection.getInstance().getDb();
  }

  public static getInstance(): PostAnalyticsService {
    if (!PostAnalyticsService.instance) {
      PostAnalyticsService.instance = new PostAnalyticsService();
    }
    return PostAnalyticsService.instance;
  }

  // ❌ PROBLEMA: Cache sem TTL e sem limite de tamanho
  async getPostMetrics(postId: number): Promise<any> {
    const cacheKey = `metrics_${postId}`;
    
    if (this.cache.has(cacheKey)) {
      return this.cache.get(cacheKey);
    }

    return new Promise((resolve, reject) => {
      const query = `
        SELECT 
          p.*,
          u.name as userName,
          COUNT(DISTINCT c.id) as commentCount,
          COUNT(DISTINCT pt.tagId) as tagCount
        FROM posts p
        LEFT JOIN users u ON p.userId = u.id
        LEFT JOIN comments c ON p.id = c.postId
        LEFT JOIN post_tags pt ON p.id = pt.postId
        WHERE p.id = ?
        GROUP BY p.id
      `;

      this.db.get(query, [postId], (err, result: any) => {
        if (err) {
          reject(err);
          return;
        }

        const metrics = {
          postId,
          title: result?.title,
          userName: result?.userName,
          commentCount: result?.commentCount || 0,
          tagCount: result?.tagCount || 0,
          createdAt: result?.createdAt,
          analyzedAt: new Date()
        };

        // ❌ PROBLEMA: Cache cresce indefinidamente
        this.cache.set(cacheKey, metrics);
        resolve(metrics);
      });
    });
  }

  // ❌ PROBLEMA: Método síncrono que deveria ser assíncrono
  calculateEngagementScore(post: Post): number {
    const daysSinceCreation = (Date.now() - new Date(post.createdAt).getTime()) / (1000 * 60 * 60 * 24);
    const commentsPerDay = (post.comments?.length || 0) / (daysSinceCreation || 1);
    const tagsWeight = (post.tags?.length || 0) * 0.5;
    
    return Math.round((commentsPerDay + tagsWeight) * 100) / 100;
  }

  // ❌ PROBLEMA: Não limpa cache periodicamente
  clearCache(): void {
    this.cache.clear();
  }
}

File: Csharp/Services/PostAnalyticsService.cs
using BlogSystem.Data;
using BlogSystem.Models;
using Microsoft.EntityFrameworkCore;

namespace BlogSystem.Services;

public class PostAnalyticsService
{
    private readonly BlogContext _context;
    private readonly ILogger<PostAnalyticsService> _logger;

    public PostAnalyticsService(BlogContext context, ILogger<PostAnalyticsService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<Post?> GetPostWithAnalyticsAsync(int postId)
    {
        _logger.LogInformation("Getting post {PostId} with analytics", postId);
        
        return await _context.Posts
            .Include(p => p.User)
            .Include(p => p.Comments)
            .ThenInclude(c => c.User)
            .Include(p => p.Tags)
            .FirstOrDefaultAsync(p => p.Id == postId);
    }

    public async Task<List<Post>> GetAllPostsWithAnalyticsAsync()
    {
        _logger.LogInformation("Getting all posts with analytics");
        
        return await _context.Posts
            .Include(p => p.User)
            .Include(p => p.Comments)
            .ThenInclude(c => c.User)
            .Include(p => p.Tags)
            .ToListAsync();
    }

    public async Task<object> GetPostMetricsAsync(int postId)
    {
        _logger.LogInformation("Calculating metrics for post {PostId}", postId);
        
        var post = await _context.Posts
            .Include(p => p.Comments)
            .Include(p => p.Tags)
            .FirstOrDefaultAsync(p => p.Id == postId);
            
        if (post == null) return null;
        
        return new
        {
            PostId = postId,
            Title = post.Title,
            CommentCount = post.Comments.Count,
            TagCount = post.Tags.Count,
            CreatedAt = post.CreatedAt,
            AnalyzedAt = DateTime.Now
        };
    }

    public async Task<object> GetUserPostStatsAsync(int userId)
    {
        _logger.LogInformation("Calculating stats for user {UserId}", userId);
        
        var posts = await _context.Posts
            .Where(p => p.UserId == userId)
            .Include(p => p.Comments)
            .Include(p => p.Tags)
            .ToListAsync();
        
        return new
        {
            UserId = userId,
            TotalPosts = posts.Count,
            TotalComments = posts.Sum(p => p.Comments.Count),
            TotalTags = posts.SelectMany(p => p.Tags).Distinct().Count(),
            AnalyzedAt = DateTime.Now
        };
    }

    public double CalculateEngagementScore(Post post)
    {
        _logger.LogDebug("Calculating engagement score for post {PostId}", post.Id);
        
        var daysSinceCreation = (DateTime.Now - post.CreatedAt).TotalDays;
        if (daysSinceCreation == 0) daysSinceCreation = 1;
        
        var commentsPerDay = post.Comments.Count / daysSinceCreation;
        var tagsWeight = post.Tags.Count * 0.5;
        
        return Math.Round(commentsPerDay + tagsWeight, 2);
    }
}
